src/Features/Core/Portable/CodeFixes/AddImport/AbstractAddImportCodeFixProvider.cs

```diff
@@ -75,12 +75,17 @@ internal abstract partial class AbstractAddImportCodeFixProvider<TSimpleNameSynt
 
                         var finder = new SymbolReferenceFinder(this, document, semanticModel, diagnostic, node, cancellationToken);
 
+                        // Caches so we don't produce the same data multiple times while searching 
+                        // all over the solution.
+                        var projectToAssembly = new ConcurrentDictionary<Project, AsyncLazy<IAssemblySymbol>>();
+                        var referenceToCompilation = new ConcurrentDictionary<PortableExecutableReference, Compilation>();
```

Comment:
Watch out for object and object[] allocations (the locking structures used for concurrency). Consider setting the maxConcurrency to a small-ish constant value (I use 2). And increase it if you see lock contention.